common/src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java

```diff
@@ -71,38 +71,54 @@ public class GTItems {
     //////////////////////////////////////
     //*****     Material Items    ******//
     //////////////////////////////////////
+
+
+
+
+    // Externally Accessible Values
     public static Table<TagPrefix, Material, ItemEntry<TagPrefixItem>> MATERIAL_ITEMS;
+    public final static Table<MaterialToolTier, GTToolType, ItemEntry<GTToolItem>> TOOL_ITEMS =
+            ArrayTable.create(GTRegistries.MATERIALS.values().stream().filter(mat -> mat.hasProperty(PropertyKey.TOOL)).map(Material::getToolTier).toList(),
+                    Arrays.stream(GTToolType.values()).toList());
 
-    public static void generateMaterialItems() {
+    // Table Builders
+    private static final ImmutableTable.Builder<TagPrefix, Material, ItemEntry<TagPrefixItem>> MATERIAL_ITEMS_BUILDER = ImmutableTable.builder();
+
+    // Material Item Generation
+    private static boolean allowedItemMaterial(TagPrefix tagPrefix, Material material) {
+        boolean doGenerateItem = tagPrefix.doGenerateItem(material);
+        return doGenerateItem;
+    }
+    private static void generateMaterialItem(TagPrefix tagPrefix, Material material) {
+        var entry = REGISTRATE.item(material.getName() + "_" + toLowerCaseUnder(tagPrefix.name), properties -> new TagPrefixItem(properties, tagPrefix, material))
+                .setData(ProviderType.LANG, NonNullBiConsumer.noop())
+                .transform(unificationItem(tagPrefix, material))
+                .properties(p -> p.stacksTo(tagPrefix.maxStackSize()))
+                .model(NonNullBiConsumer.noop())
+                .color(() -> TagPrefixItem::tintColor)
+                .register();
+        MATERIAL_ITEMS_BUILDER.put(tagPrefix, material, entry);
+    }
+    private static void generateMaterialItems() {
         REGISTRATE.creativeModeTab(() -> MATERIAL_ITEM);
-        ImmutableTable.Builder<TagPrefix, Material, ItemEntry<TagPrefixItem>> builder = ImmutableTable.builder();
         for (var tagPrefix : TagPrefix.values()) {
-            if (tagPrefix.doGenerateItem()) {
```

Comment:
you never check this, some tagprefix don't have any items and using this would remove a lot of extra looping.

Reply:
Sounds good, will be reimplemented.